core/src/main/scala/org/broadinstitute/dsde/rawls/deltalayer/DeltaLayerTranslator.scala

```diff
@@ -0,0 +1,132 @@
+package org.broadinstitute.dsde.rawls.deltalayer
+
+import akka.http.scaladsl.model.StatusCodes.{BadRequest, InternalServerError}
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.rawls.model.{Attribute, AttributeBoolean, AttributeFormat, AttributeName, AttributeNull, AttributeNumber, AttributeString, AttributeValueEmptyList, AttributeValueList, JsonSupport, PlainArrayAttributeListSerializer}
+import org.broadinstitute.dsde.rawls.model.AttributeName.toDelimitedName
+import org.broadinstitute.dsde.rawls.model.AttributeUpdateOperations.{AddUpdateAttribute, AttributeUpdateOperation, EntityUpdateDefinition}
+import org.broadinstitute.dsde.rawls.model.deltalayer.v1.DeltaRow
+
+import java.util.UUID
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Utilities to translate Rawls' batchUpsert JSON/model classes into Delta Layer's JSON/model classes
+ */
+object DeltaLayerTranslator extends JsonSupport with LazyLogging {
+
+  implicit override val attributeFormat = new AttributeFormat with PlainArrayAttributeListSerializer
+
+  val ERR_INVALID_ENTITYNAME = "This method only accepts entity names that are valid UUIDs, e.g. datarepo_row_id"
+  val ERR_INVALID_OPERATION = "This method only accepts AddUpdateAttribute operations."
+  val ERR_EMPTY_OPERATIONS = "This method requires at least one AddUpdateAttribute operation."
+  val ERR_INVALID_DATATYPE = "This method only accepts string, boolean, and numeric scalar values."
+  val ERR_EMPTY_DATATYPE = "This method requires at least one string, boolean, or numeric scalar value to update."
+
+  /**
+   * Inspects the incoming EntityUpdateDefinitions and throws exceptions if any of those updates
+   * are illegal.
+   * @param entityUpdates the updates to validate
+   * @return the original set of updates, if all are valid
+   */
+  def validateEntityUpdates(entityUpdates: Seq[EntityUpdateDefinition]): Seq[EntityUpdateDefinition] = {
+    // verify that all entity names are legal UUIDs - they should be datarepo_row_ids
```

Comment:
I think there's a bunch of mapping, partitioning and collecting that's going on her that could maybe be combined? For example, you're doing `allOps.partition` twice and getting all the AddUpdateAttribute operations each time.

Reply:
looking …

Reply:
ok, refactored this, hopefully it's better now!